src/optimizer/filter_pushdown.cpp

```diff
@@ -1,17 +1,89 @@
-#include "duckdb/optimizer/optimizer.hpp"
+#include "duckdb/planner/operator/logical_projection.hpp"
+#include "duckdb/planner/operator/logical_comparison_join.hpp"
 
 namespace duckdb {
 
 using Filter = FilterPushdown::Filter;
 
-FilterPushdown::FilterPushdown(Optimizer &optimizer, bool convert_mark_joins)
+static unordered_set<idx_t> GetMarkJoinIndexes(LogicalOperator &plan, unordered_set<idx_t> &table_bindings) {
+	unordered_set<idx_t> projected_mark_join_indexes;
+	switch (plan.type) {
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN: {
+		auto &join = plan.Cast<LogicalComparisonJoin>();
+		if (join.join_type != JoinType::MARK) {
+			break;
+		}
+		// if the projected table bindings include the mark join index,
+		if (table_bindings.find(join.mark_index) != table_bindings.end()) {
+			projected_mark_join_indexes.insert(join.mark_index);
+		}
+		break;
+	}
+	// you need to store table.column index.
+	// if you get to a projection, you need to change the table_bindings passed so they reflect the
+	// table index of the original expression they originated from.
+	case LogicalOperatorType::LOGICAL_PROJECTION: {
+		// when we encounter a projection, replace the table_bindings with
+		// the tables in the projection
+		auto plan_bindings = plan.GetColumnBindings();
+		auto &proj = plan.Cast<LogicalProjection>();
+		auto proj_bindings = proj.GetColumnBindings();
+		unordered_set<idx_t> new_table_bindings;
+		for (auto &binding : proj_bindings) {
+			auto col_index = binding.column_index;
+			auto &expr = proj.expressions.at(col_index);
+			vector<ColumnBinding> bindings_to_keep;
+			ExpressionIterator::EnumerateExpression(expr, [&](Expression &child) {
+				if (expr->expression_class == ExpressionClass::BOUND_COLUMN_REF) {
+					auto &col_ref = expr->Cast<BoundColumnRefExpression>();
+					bindings_to_keep.push_back(col_ref.binding);
+				}
+			});
+			for (auto &expr_binding : bindings_to_keep) {
+				new_table_bindings.insert(expr_binding.table_index);
+			}
+			table_bindings = new_table_bindings;
+		}
+		break;
+	}
+	default:
+		break;
+	}
+
+	// recurse into the children to find mark joins and project their indexes.
+	for (auto &child : plan.children) {
+		auto extra_mark_indexes = GetMarkJoinIndexes(*child, table_bindings);
+		for (auto extra_index : extra_mark_indexes) {
+			projected_mark_join_indexes.insert(extra_index);
+		}
+	}
+	return projected_mark_join_indexes;
+}
+
+FilterPushdown::FilterPushdown(Optimizer &optimizer, LogicalOperator &plan, bool convert_mark_joins)
     : optimizer(optimizer), combiner(optimizer.context), convert_mark_joins(convert_mark_joins) {
+
+	unordered_set<idx_t> table_bindings;
+	// other operators have not yet removed unused columns, so they will project mark joins
+	// even though they are not needed at this time
+	for (auto &binding : plan.GetColumnBindings()) {
+		table_bindings.insert(binding.table_index);
+	}
+	projected_mark_indexes = GetMarkJoinIndexes(plan, table_bindings);
```

Comment:
We are now calling `GetMarkJoinIndexes` for every single `FilterPushdown` construction - when the presence of mark joins in a plan is very rare. 

I think we should invert this in a stand-alone optimizer for the mark join, i.e.: first we find a mark join, then after finding a mark join, we traverse the relevant part of the query tree (e.g. `PROJECTION -> FILTER -> MARK JOIN`, or `AGGREGATE -> FILTER -> MARK`) to look for occurrences of the marker.

Then if we **only** find it in the filter we turn the mark join into a semi join.

If no mark joins are present we don't need to do any additional work in that case.

Reply:
I agree that this should probably just be a different operator, but I don't think `GetMarkJoinIndexes` is called for every construction. It should just be called once when FilterPushdown is started then passed down to child constructions. The extra constructor here `filter_pushdown.hpp:22` should be used, but I can see I missed a few cases. 

Regardless, I'll work on moving this functionality to a new optimizer